Support scrub to zoom for alternate charts

lovincyrus opened the pull request:
Closes #4857

- [x] enable interval selection
- [x] how to capture the interval values?
- [x] bind the intervals to `$dashboardStore?.selectedScrubRange?.start and $dashboardStore?.selectedScrubRange?.end`
- [x] `zoomScrub` based on the scrub range then `setSelectedTimeRange`
- [x] enable brush for stacked area and stacked bar
- [x] clear interval on esc
- [x] only set scrub range on drag end 
- [x] explore [full height hover on bar chart](https://vega.github.io/editor/#/examples/vega-lite/interactive_histogram_full_height_hover)
- [x] access brushend
- [x] check isScrubbing for alt charts (blocked by the ability to access brushend)
- [x] fix band size
- [x] check bar band relative size configs in vega and vega lite specs
- [x] update bars style that are in the selected interval
- [x] disable tooltip for the scrubbed interval range (configurable via global bar tooltip config)
- [x] check vega chart theme
- [x] unable to scrub bar chart when comparison is enabled
- [x] esc to clear the selection interval
- [x] revert full height hover support, can land scrub to zoom first
- [x] fix updateVegaOnTableHover unrecognized "hover_tuple_fields"
- [x] clear the interval even when switching chart type
- [x] configure other chart types' color based on brush param
- [x] to prune diff changed
- [x] fix WARN Config.customFormatTypes is not true, thus custom format type and format for channel y are dropped.
- [x] remove tooltip when isScrubbing https://github.com/rilldata/rill/pull/5483#issuecomment-2330238319
- [x] fix precision of scrub range https://github.com/rilldata/rill/pull/5483#issuecomment-2330244184
- [x] disable tooltip when scrubbing
- [x] fix https://github.com/rilldata/rill/pull/5483#issuecomment-2335121414

Refs https://github.com/vega/vega-lite/issues/3338, https://github.com/vega/vega-lite/issues/5341, https://github.com/vega/vega-lite/releases/tag/v5.0.0

Update 1: Vega-Lite does not offer direct access to the signals API. To handle events, we would need to compile the Vega-Lite spec into Vega and then add custom signals. To simplify event handling, I’ve opted to use JavaScript to dispatch and manage changes. See: https://github.com/rilldata/rill/pull/5483/commits/59b42ea510c317b3af0138d54ff200d3d5aa3b00

Update 2: Compile vega lite spec → vega spec and add vega signals to configure brushend so we can configure the `isScrubbing` state. https://github.com/rilldata/rill/pull/5483/commits/74471034d8a8afdb11f61ed3011aff8ff71a77de

Update 3: Fun fact: Vega spec doesn't support relative bar width band size.

> In Vega-Lite, bar widths can be adjusted using the bandSize property for banded scales, which effectively controls the width of bars in a banded scale. However, Vega itself does not provide a direct bandSize equivalent. Instead, Vega relies on a combination of encoding channels and layout configurations to control bar widths.

> In Vega, you can use scale with domain and range to control the width of bars, but you might need to calculate and set these values dynamically if you need relative sizing. You can also use mark properties and layout settings to manage the width and spacing of bars.

> So, while Vega-Lite provides a convenient way to handle relative bar widths with bandSize, Vega requires a more manual approach to achieve similar results.

Update 4:

https://github.com/user-attachments/assets/0bd8baa1-5c0a-4be6-b316-5ca1a123493a

Update 5: Keeping vega lite → vega spec changes to support more advanced interactions down the road

djbarnwal commented:
Did an initial UXQA pass

1. Default bars look a bit lighter than they are on main
![image](https://github.com/user-attachments/assets/d76cd93a-7739-40ae-a95f-eef5ee90a125)

2. The color of the bars when comparing with time is now different
![image](https://github.com/user-attachments/assets/0595dede-61f9-455e-b99b-827d2d48cb8c)

3. The zoom range and the colored bar range is not the same (8AM should be inclusive and 3PM should be exclusive)
![image](https://github.com/user-attachments/assets/8895038e-db55-4160-a2d3-0ff624899795)

4. Can't see tooltip on the bars which are inside the scrub range.

lovincyrus commented:
> Did an initial UXQA pass
> 
> 1. Default bars look a bit lighter than they are on main
>    ![image](https://private-user-images.githubusercontent.com/4402679/363950539-d76cd93a-7739-40ae-a95f-eef5ee90a125.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjU0MDMyMTksIm5iZiI6MTcyNTQwMjkxOSwicGF0aCI6Ii80NDAyNjc5LzM2Mzk1MDUzOS1kNzZjZDkzYS03NzM5LTQwYWUtYTk1Zi1lZWY1ZWU5MGExMjUucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDkwMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA5MDNUMjIzNTE5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9Yjg0NTNhNTcwMzBjY2U3NTU1MTc4NzBjMjJmYjQ2MGY3OWIwOWM3MzYyNzYyNGQwYTgyMjNhODM2Yjc3OGNiNCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.NUpIfYKzaFM0JVgU5AOPVRZfi8oKYvcuXmHSGSQMQxg)
> 2. The color of the bars when comparing with time is now different
>    ![image](https://private-user-images.githubusercontent.com/4402679/363950772-0595dede-61f9-455e-b99b-827d2d48cb8c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjU0MDMyMTksIm5iZiI6MTcyNTQwMjkxOSwicGF0aCI6Ii80NDAyNjc5LzM2Mzk1MDc3Mi0wNTk1ZGVkZS02MWY5LTQ1NWUtYjk5Yi04MjdkMmQ0OGNiOGMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDkwMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA5MDNUMjIzNTE5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NWRmMzgzYTNiMTczZWQwNjBlZjFhNzk1ZTRhNzdiZmQxZTRhOWM0ZDQ1NTliMjRkM2NjMjllMjY1OTJmMmVjYiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.QXb99efu9LcFgipcW6xdeBu9wD2WaOrWIF3wtPCciBA)
> 3. The zoom range and the colored bar range is not the same (8AM should be inclusive and 3PM should be exclusive)
>    ![image](https://private-user-images.githubusercontent.com/4402679/363994524-8895038e-db55-4160-a2d3-0ff624899795.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjU0MDMyMTksIm5iZiI6MTcyNTQwMjkxOSwicGF0aCI6Ii80NDAyNjc5LzM2Mzk5NDUyNC04ODk1MDM4ZS1kYjU1LTQxNjAtYTJkMy0wZmY2MjQ4OTk3OTUucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDkwMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA5MDNUMjIzNTE5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NmI2YjRiMjc3MDk4NjlkNzQ5NDQzZmY1OTJmMDI4MWZiNDA3YmYxMzg1YmFmMGZkZDc0MmQwOTIxYTA3OWViYyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.I6W3J05QGb0Hewt1x9ECDrOAmPiOciJAygzdNLFXAU0)
> 4. Can't see tooltip on the bars which are inside the scrub range.

Fixed 1 and 2.

3 - Looks like the entire scrub range is being taken into consideration rather than just the highlighted bars. This provides more granular controls since we can even start scrubbing from the whitespace.

https://github.com/user-attachments/assets/362feaa5-9407-4ee7-999d-1b91299d2348

The current measure chart's behavior is snapping to a fixed point. Do you think we should filter out the highlighted values within the scrub range or keep the current behavior? @djbarnwal 

4 - This appears to be a known limitation of vega library when hovering over the brush selection layer. https://github.com/vega/vega-lite/issues/6003 We can explore workarounds in the follow up if needed.

djbarnwal commented:
Also, we should disable tooltips when scrubbing, otherwise we see a tooltip as we scrub. Also at the end of scrub the tooltip for the last bar still remains

lovincyrus commented:
> Also, we should disable tooltips when scrubbing, otherwise we see a tooltip as we scrub.

Absolutely! I tried to disable the tooltip during scrubbing, but unfortunately, it didn't work. Here's the past commit: https://github.com/rilldata/rill/pull/5483/commits/20ec4c221d5f3648e9400a9df819943cd0f2997d

https://github.com/vega/vega-lite/issues/8750
https://vega.github.io/vega-lite/docs/tooltip.html#disable-tooltips
https://github.com/vega/altair/issues/3196

```
    tooltip: [
      {
        condition: {
          test: "!length(data('brush_store'))",
          field: timeField.tooltipName ? timeField.tooltipName : timeField.name,
          type: "temporal",
          title: "Time",
          format: "%b %d, %Y %H:%M",
        },
      },
      {
        condition: {
          test: "!length(data('brush_store'))",
          title: quantitativeField.label,
          field: quantitativeField.name,
          type: "quantitative",
          formatType: quantitativeField.formatterFunction || "number",
        },
      },
    ],
```

```
tooltip: {
      condition: {
        param: "brush",
        value: [
          {
            field: timeField.tooltipName
              ? timeField.tooltipName
              : timeField.name,
            type: "temporal",
            title: "Time",
            format: "%b %d, %Y %H:%M",
          },
          {
            title: quantitativeField.label,
            field: quantitativeField.name,
            type: "quantitative",
            formatType: quantitativeField.formatterFunction || "number",
          },
        ],
      },
      value: null,
    },
```

```
    tooltip: isScrubbing
      ? undefined
      : [
          {
            field: timeField.tooltipName
              ? timeField.tooltipName
              : timeField.name,
            type: "temporal",
            title: "Time",
            format: "%b %d, %Y %H:%M",
          },
          {
            title: quantitativeField.label,
            field: quantitativeField.name,
            type: "quantitative",
            formatType: quantitativeField.formatterFunction || "number",
          },
        ],
```

Setting tooltip to false in vega config also didn't work.
```
  mark: {
    tooltip: isScrubbing ? false : undefined,
  },
```

If we want, we could technically hide the tooltip visually as a workaround as well
```
  :global(#rill-vg-tooltip) {
    @apply hidden;
  }
```

lovincyrus commented:
> For alternate charts, it's a bit tricky. For our current line charts, the points are plotted in the center so we can snap on the edges of the bucketed period. For vega charts, the bars are placed just below the axis labels.

Indeed! The source of truth of the vega chart is the interval.

![CleanShot 2024-09-04 at 15 10 10@2x](https://github.com/user-attachments/assets/16cd58cb-bc27-447a-a951-f25b8eef3080)

We should ensure that the scrubbed range is accurately reflected in the highlighted cells.﻿

![CleanShot 2024-09-04 at 14 26 11@2x](https://github.com/user-attachments/assets/759a31a8-df4f-4e72-9748-8bddb742a8f7)

lovincyrus commented:
> > Also, we should disable tooltips when scrubbing, otherwise we see a tooltip as we scrub.
> 
> Absolutely! I tried to disable the tooltip during scrubbing, but unfortunately, it didn't work. Here's the past commit: [20ec4c2](https://github.com/rilldata/rill/commit/20ec4c221d5f3648e9400a9df819943cd0f2997d)
> 
> [vega/vega-lite#8750](https://github.com/vega/vega-lite/issues/8750) https://vega.github.io/vega-lite/docs/tooltip.html#disable-tooltips [vega/altair#3196](https://github.com/vega/altair/issues/3196)
> 
> ```
>     tooltip: [
>       {
>         condition: {
>           test: "!length(data('brush_store'))",
>           field: timeField.tooltipName ? timeField.tooltipName : timeField.name,
>           type: "temporal",
>           title: "Time",
>           format: "%b %d, %Y %H:%M",
>         },
>       },
>       {
>         condition: {
>           test: "!length(data('brush_store'))",
>           title: quantitativeField.label,
>           field: quantitativeField.name,
>           type: "quantitative",
>           formatType: quantitativeField.formatterFunction || "number",
>         },
>       },
>     ],
> ```
> 
> ```
> tooltip: {
>       condition: {
>         param: "brush",
>         value: [
>           {
>             field: timeField.tooltipName
>               ? timeField.tooltipName
>               : timeField.name,
>             type: "temporal",
>             title: "Time",
>             format: "%b %d, %Y %H:%M",
>           },
>           {
>             title: quantitativeField.label,
>             field: quantitativeField.name,
>             type: "quantitative",
>             formatType: quantitativeField.formatterFunction || "number",
>           },
>         ],
>       },
>       value: null,
>     },
> ```
> 
> ```
>     tooltip: isScrubbing
>       ? undefined
>       : [
>           {
>             field: timeField.tooltipName
>               ? timeField.tooltipName
>               : timeField.name,
>             type: "temporal",
>             title: "Time",
>             format: "%b %d, %Y %H:%M",
>           },
>           {
>             title: quantitativeField.label,
>             field: quantitativeField.name,
>             type: "quantitative",
>             formatType: quantitativeField.formatterFunction || "number",
>           },
>         ],
> ```
> 
> Setting tooltip to false in vega config also didn't work.
> 
> ```
>   mark: {
>     tooltip: isScrubbing ? false : undefined,
>   },
> ```
> 
> If we want, we could programmatically hide the tooltip visually as a workaround as well
> 
> ```
>   :global(#rill-vg-tooltip) {
>     @apply hidden;
>   }
> ```

Okay, fixed this with https://github.com/rilldata/rill/pull/5483/commits/a22c942382b1fb704aa5799682a7976807884cae by configuring the tooltip handler directly.

https://github.com/user-attachments/assets/c944065a-f0d9-4f8e-a279-439030b1214f

lovincyrus commented:
Addressed the feedback, @djbarnwal 

I think this is a good baseline. I have created some follow up tickets to keep track of the remaining issues so we can land them incrementally.

https://github.com/rilldata/rill/issues/5633, https://github.com/rilldata/rill/issues/5634

djbarnwal commented:
Have mentioned a few small but critical bugs in this loom video 

https://www.loom.com/share/62bc766c466847608ced7c1eee604860?sid=dc47bbad-77a3-40d2-bfd2-49214aabd1d1

djbarnwal commented on web-common/src/features/charts/render/vega-config.ts:
```diff
@@ -16,9 +16,6 @@ export const getRillTheme: () => Config = () => ({
   autosize: {
     type: "fit-x",
   },
-  mark: {
-    tooltip: true,
-  },
```
What's the idea behind removing this? We also use the same config for our custom dashboard charts so this removes the default tooltips from those charts.

lovincyrus replied:
> What's the idea behind removing this?

Keeping this in the global config will trigger an empty tooltip on the brush interval.

```
  mark: {
    tooltip: true,
  },
```


https://github.com/user-attachments/assets/cdda169d-bd64-44a7-bb9a-b55070425edf

lovincyrus replied:
If the global tooltip config is absolutely needed, I can enable the tooltip flag for customDashboard.

lovincyrus replied:
https://github.com/rilldata/rill/pull/5483/commits/86805de59ca62a3ccfeb317ad52e8eac113dec4b

djbarnwal commented on web-common/src/features/dashboards/time-dimension-details/charts/TDDAlternateChart.svelte:
```diff
@@ -34,6 +39,7 @@
   export let isTimeComparison: boolean;
 
   let viewVL: View;
+  let vegaSpec: any;
```
Can we add the type for Vega Spec here? (should be `VisualizationSpec` or `VgSpec`)

djbarnwal commented on web-common/src/features/dashboards/time-dimension-details/charts/utils.ts:
```diff
@@ -132,11 +137,46 @@ export function updateVegaOnTableHover(
+}
+
+// Check if vega lite spec has brush params
+// If so, we need toompile vega lite spec to vega spec
```
nit: typo

lovincyrus commented:
> Have mentioned a few small but critical bugs in this loom video
> 
> https://www.loom.com/share/62bc766c466847608ced7c1eee604860?sid=dc47bbad-77a3-40d2-bfd2-49214aabd1d1

Nice catch on the timezone inconsistencies! I'll review if any of the code paths affected these behaviors.

lovincyrus commented:
> > Have mentioned a few small but critical bugs in this loom video
> > https://www.loom.com/share/62bc766c466847608ced7c1eee604860?sid=dc47bbad-77a3-40d2-bfd2-49214aabd1d1
> 
> Nice catch on the timezone inconsistencies! I'll review if any of the code paths affected these behaviors.

Looking back at the scrub behaviors of the granular details of time range, we weren't able to scrub the chart at all when the time grain is `TIME_GRAIN_HOUR`

![CleanShot 2024-09-09 at 09 36 42@2x](https://github.com/user-attachments/assets/a6aad0dd-2bb3-458b-99e8-bb68a27c0537)

https://github.com/user-attachments/assets/cd946e96-ef9d-4091-ae29-c14fbda5c6f9

lovincyrus commented:
Fixed timezone inconsistency in the alt charts

https://github.com/user-attachments/assets/36c5e587-ccae-427b-b549-6574be32044f

![CleanShot 2024-09-09 at 17 29 54@2x](https://github.com/user-attachments/assets/84666942-f8f6-4b54-8560-24149569bebd)

djbarnwal commented on web-common/src/features/dashboards/time-dimension-details/charts/TDDAlternateChart.svelte:
```diff
@@ -117,13 +211,26 @@
+    {signalListeners}
+    {expressionFunctions}
+    {tooltipFormatter}
+    canvasDashboard={$customDashboards}
```
This should be false and not rely on the feature flag.

djbarnwal commented on web-common/src/features/dashboards/time-dimension-details/charts/TDDAlternateChart.svelte:
```diff
@@ -117,13 +211,26 @@
     {signalListeners}
     {expressionFunctions}
     {tooltipFormatter}
+    canvasDashboard={$customDashboards}
```
This should be false and not rely on the feature flag.

ericpgreen2 commented on web-common/src/features/canvas-components/render/VegaRenderer.svelte:
```diff
@@ -0,0 +1,153 @@
+  export let signalListeners: SignalListeners = {};
+  export let expressionFunctions: ExpressionFunction = {};
+  export let error: string | null = null;
+  export let canvasDashboard = false;
+  export let chartView = false;
```
It seems a little off for a "Vega[Lite]Renderer" to be aware of canvas dashboards. I suppose I'd expect that  a "Vega[Lite]Renderer" would solely render Vega visualizations, and any Rill application logic (e.g. references to canvases or the runtime) would be handled outside of it.

(Though I noticed that the `VegaLiteRenderer` also includes a `canvasDashboard` prop, so I do agree with keeping the two Vega renderers symmetrical.)

lovincyrus replied:
https://rilldata.slack.com/archives/C04AGHL77PS/p1726514541389759?thread_ts=1726509359.103919&cid=C04AGHL77PS

djbarnwal reviewed (approved):
Thanks for the PR!


A bug I noticed while playing around was that for some scrubbbed ranges, the table incorrectly shows no data. We can address this in a follow up burndown list.

![image](https://github.com/user-attachments/assets/9b5a5b99-95eb-40b1-af05-2a621d3f1739)
